tests/framework/randomness/test_crn.py

```diff
@@ -0,0 +1,301 @@
+        pytest.param(SequentialPopulation, False),
+    ],
+)
+def test_multi_sim_reproducibility_with_different_pop_growth(with_crn, pop_class):
+    if with_crn:
+        configuration = {"randomness": {"key_columns": ["crn_attr1", "crn_attr2"]}}
+    else:
+        configuration = {}
+
+    short, long = 1, 3
+    sim1 = InteractiveContext(
+        components=[pop_class(with_crn=with_crn, sims_to_add=cycle([short, short]))],
+        configuration=configuration,
+    )
+    sim2 = InteractiveContext(
+        components=[pop_class(with_crn=with_crn, sims_to_add=cycle([long, long]))],
+        configuration=configuration,
+    )
+
+    pop1 = sim1.get_population().set_index(["crn_attr1", "crn_attr2"])
+    pop2 = sim2.get_population().set_index(["crn_attr1", "crn_attr2"])
+    initial_pop_size = len(pop1)
+    assert_frame_equal(pop1, pop2)
+
+    time_steps = 7
+    for i in range(time_steps):
+        sim1.step()
+        sim2.step()
+
+    pop1 = sim1.get_population().set_index(["crn_attr1", "crn_attr2"]).drop(columns="tracked")
+    pop2 = sim2.get_population().set_index(["crn_attr1", "crn_attr2"]).drop(columns="tracked")
+
+    if with_crn:
+        overlap = pop1.index.intersection(pop2.index)
+        assert len(overlap) > initial_pop_size
+        assert_frame_equal(pop1.loc[overlap], pop2.loc[overlap])
+    else:
+        overlap = pop1.index[:initial_pop_size]
+        assert_frame_equal(pop1.loc[overlap], pop2.loc[overlap])
+
+
+class BrokenPopulation(BasePopulation):
+    """CRN system falls over if the first CRN attribute is an int or float."""
+
+    def on_initialize_simulants(self, pop_data: SimulantData):
+        crn_attr = (1_000_000 * self.randomness_init.get_draw(index=pop_data.index)).astype(
+            int
+        )
+        population = pd.DataFrame(
+            {"crn_attr1": crn_attr, "crn_attr2": pop_data.creation_time},
+            index=pop_data.index,
+        )
+
+        if self.with_crn:
+            self.register(population)
+
+        population["other_attr1"] = self.randomness_other.get_draw(
+            pop_data.index,
+            additional_key="attr1",
+        )
+        self.population_view.update(population)
+
+
+@pytest.mark.parametrize(
+    "with_crn, sims_to_add",
+    [
+        pytest.param(True, cycle([0]), marks=pytest.mark.xfail),
+        pytest.param(True, cycle([1]), marks=pytest.mark.xfail),
+        pytest.param(False, cycle([0])),
+        pytest.param(False, cycle([1])),
+    ],
+)
+def test_failure_path_when_first_crn_attribute_not_datelike(with_crn, sims_to_add):
+    if with_crn:
+        configuration = {"randomness": {"key_columns": ["crn_attr1", "crn_attr2"]}}
+    else:
+        configuration = {}
+
+    sim = InteractiveContext(
+        components=[BrokenPopulation(with_crn=with_crn, sims_to_add=sims_to_add)],
+        configuration=configuration,
+    )
+
+    for i in range(5):
+        sim.step()
```

Comment:
How does this test pass? I was expecting an `assert` or `pytest.raises`

Reply:
This test fails when the `with_crn` flag is true. But it's marked as `pytest.mark.xfail`, i.e., an expected failure.  `pytest.raises` is the right tool when you are testing error paths in your code, e.g., malformed inputs to a function.  `pytest.mark.xfail` is useful to highlight a bug in the code without an existing patch.  It's essentially a regression test I'm putting in ahead of time.